<a href="https://colab.research.google.com/github/vilar18/Hello-Python/blob/main/PAC_2_Anal%C3%ADtica_de_clients_Aula_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#PAC 2 - Analítica de Clients 2024-2025

##Dani Vilar (vilar18)


In [2]:
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


##Llibreries i carrega de dades

In [3]:
# Llibreries
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots
import plotly.io as pio


In [4]:
file_path = '/content/drive/MyDrive/dades/UOC/Analítica de clients/data_r2_25.csv'
df = pd.read_csv(file_path, sep=',')
df.head()


CUSTOMER_ID  SALDO_CUENTA  FRECUENCIA_ACTUALIZACION_SALDO  IMPORTE_COMPRAS  \
0      C18658      2.850024                        0.181818           114.56   
1      C18844    469.319087                        0.818182           383.81   
2      C12288   1504.334228                        0.909091           565.00   
3      C12767   2061.835735                        1.000000             0.00   
4      C11595     18.218714                        0.272727            89.00   

   IMPORTE_COMPRAS_ONOFF  IMPORTE_COMPRAS_APLAZO  IMPORTE_EFECTIVO_ADELANTADO  \
0                    0.0                  114.56                     0.000000   
1                    0.0                  383.81                   870.533909   
2                  565.0                    0.00                     0.000000   
3                    0.0                    0.00                  4320.771337   
4                    0.0                   89.00                     0.000000   

   FRECUENCIA_COMPRAS  FRECUENCIA_COMPRAS_APLAZO  \
0            0.250000                   0.250000   
1            0.583333                   0.500000   
2            0.250000                   0.000000   
3            0.000000                   0.000000   
4            0.083333                   0.083333   

   FRECUENCIA_EFECTIVO_ADELANTADO  NUM_TRANSACCIONES_EFECTIVO_ADELANTADO  \
0                        0.000000                                      0   
1                        0.333333                                      6   
2                        0.000000                                      0   
3                        0.500000                                     12   
4                        0.000000                                      0   

   NUM_TRANSACCIONES_COMPRAS  LIMITE_CREDITO  IMPORTE_PAGOS  \
0                          3          1000.0     403.212926   
1                         14          1500.0      57.924613   
2                          3          2200.0    2511.315474   
3                          0          4000.0    5451.599084   
4                          1          4000.0    1095.921001   

   IMPORTE_MINIMO_PAGOS  TOTAL_TARJ_CREDITO  TOTAL_VISITAS_ONLINE  \
0            130.700011                  12                     3   
1            335.598994                  12                    18   
2            471.295665                  12                    20   
3            556.327998                  12                    19   
4            173.068884                  12                     4   

   TOTAL_VISITAS_OFICINA  GENDER    JOB_SECTOR  
0                      0    MALE   Engineering  
1                      0  FEMALE       Finance  
2                      5    MALE       Finance  
3                      0    MALE  Construction  
4                      1    MALE        Retail

##**Pregunta 1**

####a. Quantes observacions i variables té el dataset? Quin tipus de variables són?

Aquest dataset conté informació sobre **7994** clients, representats com a observacions (files), cadascun descrit per **20 variables** (columnes).

La majoria de les variables són de naturalesa numèrica. Hi ha **12 variables de tipus float64**, que corresponen a mètriques contínues com saldos (SALDO_CUENTA), imports monetaris (IMPORTE_COMPRAS, LIMITE_CREDITO) i freqüències (FRECUENCIA_...). A més, hi ha **5 variables de tipus int64**, que representen recomptes discrets, com el nombre de transaccions (NUM_TRANSACCIONES_...), targetes totals (TOTAL_TARJ_CREDITO) o visites (TOTAL_VISITAS_...).

Les **3 variables restants són de tipus object**, generalment utilitzades per a dades textuals o categòriques. Aquestes inclouen l'identificador únic del client (CUSTOMER_ID) i característiques descriptives com el gènere (GENDER) i el sector laboral (JOB_SECTOR).
És important notar que gairebé totes les columnes tenen dades completes per a les 7994 observacions, excepte LIMITE_CREDITO que li falta un valor, i IMPORTE_MINIMO_PAGOS, que presenta 283 valors absents.



In [5]:
num_observacions = df.shape[0]
num_variables = df.shape[1]

print(f"El dataset té:")
print(f"- {num_observacions} observacions (files)")
print(f"- {num_variables} variables (columnes)")
print("-" * 30) # Separador

# Quin tipus de variables són?
print("Tipus de variables (columnes):")

df.info()

El dataset té:
- 7994 observacions (files)
- 20 variables (columnes)
------------------------------
Tipus de variables (columnes):
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7994 entries, 0 to 7993
Data columns (total 20 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   CUSTOMER_ID                            7994 non-null   object 
 1   SALDO_CUENTA                           7994 non-null   float64
 2   FRECUENCIA_ACTUALIZACION_SALDO         7994 non-null   float64
 3   IMPORTE_COMPRAS                        7994 non-null   float64
 4   IMPORTE_COMPRAS_ONOFF                  7994 non-null   float64
 5   IMPORTE_COMPRAS_APLAZO                 7994 non-null   float64
 6   IMPORTE_EFECTIVO_ADELANTADO            7994 non-null   float64
 7   FRECUENCIA_COMPRAS                     7994 non-null   float64
 8   FRECUENCIA_COMPRAS_APLAZO              7994 non-null   float64
 9   FRECUENCI

####b. Té el dataset sesgues de gènere?

El dataset conté dades de 7994 clients, amb una distribució de gènere lleugerament desequilibrada: **55.2% de les observacions corresponen a dones (FEMALE) i 44.8% a homes (MALE)**. Aquesta diferència d'aproximadament 10 punts percentuals indica una sobrerepresentació moderada de dones en la mostra. Tot i que no és un desequilibri extrem, és un factor a tenir en compte, ja que models entrenats amb aquestes dades podrien, potencialment, generalitzar lleugerament millor per al grup majoritari si no s'apliquen tècniques de mitigació.

Les mètriques numèriques clau no presenten diferències significatives que suggereixin un biaix sistèmic clar en les variables financeres principals. Per exemple, el SALDO_CUENTA mitjà i la mediana són molt similars entre gèneres (mitjana: 1566 vs 1521; mediana: 888 vs 842). El mateix passa amb l'IMPORTE_COMPRAS (mitjana: 999 vs 957; mediana: 360 vs 349) i, de forma destacada, amb el LIMITE_CREDITO, on la mitjana és gairebé idèntica (4484 vs 4460) i la mediana és exactament la mateixa (3000) per a ambdós grups. Això suggereix que, **almenys a nivell central de la distribució, no hi ha una diferència significativa en el límit de crèdit assignat** basat únicament en el gènere registrat.

Les lleugeres diferències observades en IMPORTE_PAGOS, TOTAL_VISITAS_ONLINE i NUM_TRANSACCIONES_COMPRAS (amb valors lleugerament superiors per a les dones) no semblen prou grans per indicar un biaix.

Amb tot això, l'anàlisi de la variable JOB_SECTOR revela una segregació de gènere molt marcada. Aquesta és l'evidència més clara d'un patró relacionat amb el gènere dins del dataset. Observem que sectors com Construcció (0% dones vs 12.6% homes), Enginyeria (3.2% vs 17.5%), Finances (4.8% vs 12.5%) i especialment Tecnologia (5.2% vs 32.6%) estan fortament dominats per homes en aquesta mostra. Contràriament, Màrqueting (37.5% dones vs 0% homes), Retail (20.5% vs 5.2%) i Sanitat (19.5% vs 10.8%) tenen una representació femenina molt més alta. Aquesta forta correlació entre gènere i sector laboral implica **que JOB_SECTOR actua com un proxy potencialment fort per a GENDER**. Si aquesta variable s'utilitza en models predictius, podria introduir o amplificar biaixos de gènere indirectament, fins i tot si el gènere mateix no s'utilitza explícitament.

Finalment, l'anàlisi dels valors nuls en LIMITE_CREDITO i IMPORTE_MINIMO_PAGOS mostra percentatges de dades faltants molt similars entre homes i dones (0% vs 0.03% per al límit de crèdit; 3.47% vs 3.63% per als pagaments mínims). Això suggereix que la **qualitat de les dades en termes d'exhaustivitat** no sembla estar esbiaixada per gènere en aquestes columnes específiques.

En **conclusió**, tot i que les mètriques financeres directes com el límit de crèdit no mostren biaixos alarmants entre gèneres, el dataset presenta un desequilibri moderat en la representació i, de manera molt significativa, una forta segregació per gènere en els sectors laborals. Aquest últim punt és el principal focus de preocupació pel que fa al biaix potencial.



In [6]:
# 1. Distribució de la variable Gènere
print("\n1. Distribució de Gènere:")
gender_counts = df['GENDER'].value_counts()
gender_percentages = df['GENDER'].value_counts(normalize=True) * 100
print("Recompte per Gènere:\n", gender_counts)
print("\nPercentatge per Gènere:\n", gender_percentages)

# Visualització amb Plotly Express
fig_gender_dist = px.bar(
    x=gender_counts.index,
    y=gender_counts.values,
    labels={'x': 'Gènere', 'y': 'Nombre de Clients'},
    title='Distribució de Gènere en el Dataset',
    text_auto=True, # Mostra els valors a sobre de les barres
    color_discrete_sequence=px.colors.qualitative.Pastel # Esquema de colors
)
fig_gender_dist.update_layout(xaxis_title="Gènere", yaxis_title="Nombre de Clients")
fig_gender_dist.show() # Mostra el gràfic

# 2. Comparació de mètriques numèriques clau per gènere
print("\n\n2. Comparació de Mètriques Numèriques Mitjanes per Gènere:")
numeric_cols_to_compare = [
    'SALDO_CUENTA', 'IMPORTE_COMPRAS', 'LIMITE_CREDITO',
    'IMPORTE_PAGOS', 'TOTAL_VISITAS_ONLINE', 'NUM_TRANSACCIONES_COMPRAS'
]
gender_comparison = df.groupby('GENDER')[numeric_cols_to_compare].agg(['mean', 'median'])
print(gender_comparison)

# Visualització comparativa (Boxplots en subplots amb Plotly)
# Variables a visualitzar als boxplots
boxplot_vars = ['SALDO_CUENTA', 'IMPORTE_COMPRAS', 'LIMITE_CREDITO']
# Títols per als subplots
subplot_titles = [f'Comparativa {var.replace("_", " ").title()} per Gènere' for var in boxplot_vars]

# Crear figura amb subplots
fig_boxplots = make_subplots(rows=1, cols=len(boxplot_vars), subplot_titles=subplot_titles)

# Afegir traces de boxplot a cada subplot
for i, var in enumerate(boxplot_vars):
    # Afegim una traça per cada gènere per poder assignar colors diferents si volem
    # O simplement deixem que Plotly ho gestioni amb x='GENDER'
    fig_boxplots.add_trace(
        go.Box(y=df[var], x=df['GENDER'], name=var, marker_color=px.colors.qualitative.Pastel[i]), # Assigna un color per variable
        row=1, col=i+1
    )
    # Actualitzem els eixos per aquest subplot específic
    fig_boxplots.update_yaxes(title_text=var.replace("_", " ").title(), row=1, col=i+1)
    fig_boxplots.update_xaxes(title_text="Gènere", row=1, col=i+1)


# Actualitzar el layout general
fig_boxplots.update_layout(
    title_text='Comparativa de Distribucions Numèriques per Gènere',
    height=500, # Ajusta l'alçada si cal
    showlegend=False # La llegenda no és molt útil aquí
)
fig_boxplots.show() # Mostra el gràfic combinat

# 3. Relació entre Gènere i altres variables categòriques (ex: JOB_SECTOR)
print("\n\n3. Relació entre Gènere i Sector Laboral (JOB_SECTOR):")
cross_tab = pd.crosstab(df['GENDER'], df['JOB_SECTOR'])
print("\nTaula de Contingència (Recomptes):\n", cross_tab)
cross_tab_norm = pd.crosstab(df['GENDER'], df['JOB_SECTOR'], normalize='index') * 100
print("\nTaula de Contingència (Percentatge per Gènere):\n", cross_tab_norm)

# Visualització amb Plotly Express (Grouped Bar Chart usant histogram)
fig_job_gender = px.histogram(
    df,
    x='JOB_SECTOR',
    color='GENDER',  # Agrupa i coloreja per gènere
    barmode='group',
    labels={'count': 'Nombre de Clients', 'JOB_SECTOR': 'Sector Laboral', 'GENDER': 'Gènere'},
    title='Distribució de Sectors Laborals per Gènere',
    category_orders={"JOB_SECTOR": sorted(df['JOB_SECTOR'].unique())}, # Opcional: ordena els sectors
    color_discrete_map={'MALE': px.colors.qualitative.Pastel[0], 'FEMALE': px.colors.qualitative.Pastel[1]} # Assigna colors
)
fig_job_gender.update_layout(
    xaxis_title="Sector Laboral",
    yaxis_title="Nombre de Clients",
    xaxis_tickangle=-45
)
fig_job_gender.show() # Mostra el gràfic

# 4. Comprovació de valors nuls per gènere (en variables on n'hi ha)
print("\n\n4. Percentatge de Valors Nuls per Gènere en Columnes Crítiques:")
cols_with_nulls = ['LIMITE_CREDITO', 'IMPORTE_MINIMO_PAGOS']
null_analysis = df.groupby('GENDER')[cols_with_nulls].apply(lambda x: x.isnull().mean() * 100)
print(null_analysis)

print("\n" + "="*55 + "\nFi de l'Anàlisi Preliminar d'Imbalance de Gènere (amb Plotly).")


1. Distribució de Gènere:
Recompte per Gènere:
 GENDER
FEMALE    4413
MALE      3581
Name: count, dtype: int64

Percentatge per Gènere:
 GENDER
FEMALE    55.203903
MALE      44.796097
Name: proportion, dtype: float64




2. Comparació de Mètriques Numèriques Mitjanes per Gènere:
       SALDO_CUENTA             IMPORTE_COMPRAS         LIMITE_CREDITO  \
               mean      median            mean  median           mean   
GENDER                                                                   
FEMALE  1566.214738  887.679317      999.413759  360.00    4484.020239   
MALE    1521.779540  841.744925      957.265834  348.78    4460.222998   

               IMPORTE_PAGOS             TOTAL_VISITAS_ONLINE         \
        median          mean      median                 mean median   
GENDER                                                                 
FEMALE  3000.0   1757.104868  858.248769            10.618627    9.0   
MALE    3000.0   1654.049236  832.140348            10.096900    8.0   

       NUM_TRANSACCIONES_COMPRAS         
                            mean median  
GENDER                                   
FEMALE                 14.722864    7.0  
MALE                   14.394583    7.0



3. Relació entre Gènere i Sector Laboral (JOB_SECTOR):

Taula de Contingència (Recomptes):
 JOB_SECTOR  Construction  Engineering  Finance  Healthcare  Manufacturing  \
GENDER                                                                      
FEMALE                 0          142      212         862            410   
MALE                 450          625      449         386            314   

JOB_SECTOR  Marketing  Retail  Technology  
GENDER                                     
FEMALE           1655     903         229  
MALE                0     188        1169  

Taula de Contingència (Percentatge per Gènere):
 JOB_SECTOR  Construction  Engineering    Finance  Healthcare  Manufacturing  \
GENDER                                                                        
FEMALE          0.000000     3.217766   4.803988   19.533197       9.290732   
MALE           12.566322    17.453225  12.538397   10.779112       8.768500   

JOB_SECTOR  Marketing     Retail  Technology  
GENDER 



4. Percentatge de Valors Nuls per Gènere en Columnes Crítiques:
        LIMITE_CREDITO  IMPORTE_MINIMO_PAGOS
GENDER                                      
FEMALE        0.000000              3.467029
MALE          0.027925              3.630271

Fi de l'Anàlisi Preliminar d'Imbalance de Gènere (amb Plotly).


####c. Quines mesures estadístiques destacaries donats els valors obtinguts per a cada variable?

Analitzant els resultats de describe(), podem destacar diverses mesures estadístiques clau per a diferents variables:

**Mesures de Tendència Central (Mitjana vs. Mediana / mean vs 50%):**

En moltes variables monetàries (SALDO_CUENTA, IMPORTE_COMPRAS, IMPORTE_COMPRAS_ONOFF, IMPORTE_COMPRAS_APLAZO, IMPORTE_EFECTIVO_ADELANTADO, IMPORTE_PAGOS, IMPORTE_MINIMO_PAGOS) i en els recomptes de transaccions (NUM_TRANSACCIONES_EFECTIVO_ADELANTADO, NUM_TRANSACCIONES_COMPRAS), la mitjana (mean) és significativament més alta que la mediana (50%). Això indica que la distribució d'aquestes variables està esbiaixada cap a la dreta, és a dir, hi ha un gran nombre de clients amb valors baixos o moderats, però un petit nombre de clients amb valors extremadament alts que "estiren" la mitjana cap amunt. La mediana és sovint un millor indicador del valor "típic" en aquests casos. LIMITE_CREDITO: També mostra aquesta asimetria (mitjana 4473 vs mediana 3000).

**Variables de Freqüència:**

FRECUENCIA_COMPRAS (mitjana 0.49 vs mediana 0.50) està més centrada, mentre que FRECUENCIA_COMPRAS_APLAZO (0.36 vs 0.17) i FRECUENCIA_EFECTIVO_ADELANTADO (0.13 vs 0.00) estan clarament esbiaixades, amb molts clients tenint freqüències baixes o zero (la mediana de 0 indica que almenys la meitat dels clients tenen freqüència 0 per a l'efectiu adelantat).

**Mesures de Dispersió (std, min, max, Quartils):**

Moltes de les variables monetàries i de recompte tenen una desviació estàndard (std) molt alta, sovint superior a la pròpia mitjana (SALDO_CUENTA, IMPORTE_COMPRAS, etc.). Això confirma la gran dispersió dels valors i la presència de valors extrems.

Els valors mínim (min) i màxim (max) revelen rangs enormes per a les variables financeres, des de 0 fins a desenes de milers en alguns casos (IMPORTE_MINIMO_PAGOS max > 61000!). Això suggereix la presència de possibles outliers o, si més no, una clientela molt heterogènia en termes de volum financer.

Concentració en valors baixos: El primer quartil (25%) és 0 o proper a 0 per a diverses variables (IMPORTE_COMPRAS_ONOFF, IMPORTE_COMPRAS_APLAZO, IMPORTE_EFECTIVO_ADELANTADO, FRECUENCIA_COMPRAS_APLAZO, FRECUENCIA_EFECTIVO_ADELANTADO, NUM_TRANSACCIONES_EFECTIVO_ADELANTADO), indicant que almenys el 25% dels clients no fan ús (o tenen valor zero) d'aquestes funcionalitats.

La FRECUENCIA_EFECTIVO_ADELANTADO té un valor màxim (max) de 1.5. Si aquesta variable representa una freqüència normalitzada (entre 0 i 1), aquest valor és anòmal i requereix una revisió. Podria ser un error de dades o una definició diferent de la variable.

TOTAL_TARJ_CREDITO té una distribució peculiar. El mínim és 6, el màxim és 12, però el 25è, 50è (mediana) i 75è percentil són tots 12. Això és molt destacable i significa que almenys el 75% dels clients tenen exactament 12 targetes, amb només una petita cua de clients amb menys targetes (entre 6 i 11).

**VAlors nuls:**

En destaca LIMITE_CREDITO que té 1 valor menys (7993 vs 7994) i, de forma més significativa, IMPORTE_MINIMO_PAGOS, que té 283 valors menys (7711 vs 7994)

**Estadístiques per a Variables Categòriques:**

En el cas de GENDER i JOB_SECTOR només es mostra la categoria més freqüent (top) i la seva freqüència (freq), donant una idea limitada de la seva distribució.

En **resum**, les estadístiques descriptives revelen un dataset amb molta variabilitat i asimetria en les mètriques financeres, concentració de valors baixos/zeros en certs comportaments, dades faltants en variables clau, una distribució molt particular per al nombre de targetes i un possible valor anòmal en una de les freqüències.

In [7]:
# Genera estadístiques descriptives per a les columnes NUMÈRIQUES
print("Estadístiques Descriptives (Només Columnes Numèriques):")
desc_numeric = df.describe()
print(desc_numeric)

print("\n" + "="*60 + "\n") # Separador

# Genera estadístiques descriptives per a TOTES les columnes
print("Estadístiques Descriptives (Totes les Columnes):")
desc_all = df.describe(include='all')
print(desc_all)

Estadístiques Descriptives (Només Columnes Numèriques):
       SALDO_CUENTA  FRECUENCIA_ACTUALIZACION_SALDO  IMPORTE_COMPRAS  \
count   7994.000000                     7994.000000      7994.000000   
mean    1546.309503                        0.877281       980.533134   
std     2056.495837                        0.236988      1954.559070   
min        0.000000                        0.000000         0.000000   
25%      126.134981                        0.888889        35.925000   
50%      864.868214                        1.000000       354.565000   
75%     2014.959018                        1.000000      1099.545000   
max    19043.138560                        1.000000     32539.780000   

       IMPORTE_COMPRAS_ONOFF  IMPORTE_COMPRAS_APLAZO  \
count            7994.000000             7994.000000   
mean              572.496469              408.337100   
std              1476.628493              896.445531   
min                 0.000000                0.000000   
25%            

####Pots suposar a partir d’aquestes mesures quines variables tenen outliers?

Sí, podem fer una suposició raonada sobre quines variables tenen més probabilitats de contenir outliers basant-nos en les estadístiques descriptives, especialment comparant la mitjana amb la mediana i observant la desviació estàndard i els valors màxims.

**SALDO_CUENTA:**
La mitjana (1546) és quasi el doble de la mediana (865). La desviació estàndard (2056) és superior a la mitjana. El valor màxim (19043) està molt allunyat del 75è percentil (2015).
Semblaria que hi ha saldos molt elevats que influeixen fortament en la mitjana.

**IMPORTE_COMPRAS:**
Mitjana (980) molt superior a la mediana (355). Desviació estàndard (1954) molt superior a la mitjana. Màxim (32540) molt lluny del 75è percentil (1100).
Probablement existeixen clients amb imports de compra extremadament alts.

**IMPORTE_COMPRAS_ONOFF:**
Diferència extrema entre mitjana (572) i mediana (32). Desviació estàndard (1477) molt més gran que la mitjana. Màxim (26547) enormement distant del 75è percentil (564).
Sembla que hi ha forta presència d'outliers amb valors molt alts.

**IMPORTE_COMPRAS_APLAZO:**
Mitjana (408) molt superior a la mediana (89). Desviació estàndard (896) superior a la mitjana. Màxim (22500) molt lluny del 75è percentil (467).
Hi hauria outliers amb valors elevats.

**IMPORTE_EFECTIVO_ADELANTADO:**
Diferència abismal entre mitjana (968) i mediana (0). Desviació estàndard (2050) més del doble de la mitjana. Màxim (29282) molt lluny del 75è percentil (1106).
Sembla que molts clients no utilitzen aquest servei (mediana 0), però uns pocs ho fan amb imports molt alts (outliers).

**FRECUENCIA_EFECTIVO_ADELANTADO:**
A més de la diferència entre mitjana (0.13) i mediana (0), el valor màxim (1.5) és particularment sospitós si s'espera que les freqüències estiguin entre 0 i 1. Aquest valor màxim és, per definició, un outlier (potencialment un error).

**NUM_TRANSACCIONES_EFECTIVO_ADELANTADO:**
Mitjana (3.2) molt superior a la mediana (0). Desviació estàndard (6.7) superior a la mitjana. Màxim (123) molt lluny del 75è percentil (4).
Semblaria que hi ha outliers amb un nombre molt alt de transaccions.

**NUM_TRANSACCIONES_COMPRAS:**
Mitjana (14.6) el doble de la mediana (7). Desviació estàndard (24.5) superior a la mitjana. Màxim (347) molt lluny del 75è percentil (17).
Outliers amb un nombre molt alt de compres.

**LIMITE_CREDITO:**
Mitjana (4473) bastant superior a la mediana (3000). Màxim (30000) molt lluny del 75è percentil (6500).
Existència de límits de crèdit molt alts (outliers).

**IMPORTE_PAGOS:**
Mitjana (1711) més del doble de la mediana (849). Desviació estàndard (2785) superior a la mitjana. Màxim (50721) enormement distant del 75è percentil (1897).
Imports de pagament extremadament alts com outliers.

**IMPORTE_MINIMO_PAGOS:**
Mitjana (854) molt superior a la mediana (310). Desviació estàndard (2265) molt superior a la mitjana. Màxim (61031) extremadament lluny del 75è percentil (815).
Valors molt alts d'imports mínims (outliers).

En conclusió, les variables relacionades amb imports monetaris i recomptes de transaccions són les candidates més clares a tenir outliers (valors excepcionalment alts), com ho suggereixen les grans diferències entre mitjanes i medianes, les altes desviacions estàndard i els valors màxims extrems. La FRECUENCIA_EFECTIVO_ADELANTADO també té un valor màxim potencialment anòmal.

####d. Crea les següents gràfiques y raona el que s’hi observa:
i. Distribució pell saldo del compte.

ii. Relació entre l’import de compres y el nombre de transaccions.

iii. Distribució de l’import de compres per sector de treball.

In [9]:
fig_dist_saldo = px.histogram(
    df,
    x="SALDO_CUENTA",       # Variable eix X
    nbins=100,              # Nombre de 'barres'
    marginal="box",         # Afegeix un diagrama de caixa a la part superior
    title="Distribució del Saldo del Compte",
    labels={"SALDO_CUENTA": "Saldo del Compte"}, # Etiqueta personalitzada per l'eix X
    opacity=0.8,
    color_discrete_sequence=px.colors.qualitative.Pastel # Esquema de colors
)

fig_dist_saldo.update_layout(
    yaxis_title="Freqüència (Nombre de Clients)",
    bargap=0.1 # Petit espai entre barres
)

# Mostra el gràfic
fig_dist_saldo.show()

El gràfic mostra la distribució del saldo dels comptes dels clients mitjançant un histograma i un diagrama de caixa (boxplot) superior.
L'observació més evident és la forta asimetria positiva (biaix cap a la dreta) de la distribució. La barra més alta de l'histograma es troba a l'extrem esquerre, indicant que la gran majoria dels clients tenen saldos molt baixos, propers a zero. A mesura que el saldo augmenta, la freqüència (nombre de clients) disminueix dràsticament.

El diagrama de caixa (boxplot) confirma aquesta observació. La caixa, que representa l'interval interquartílic (IQR) on es troba el 50% central dels clients, és relativament estreta i està situada a la part baixa de l'escala. La línia dins la caixa (la mediana) també es troba en un valor baix de saldo.

A més, el boxplot revela la presència de nombrosos valors atípics (outliers), representats pels punts individuals que s'estenen molt cap a la dreta, més enllà del bigoti superior. Tot i que la majoria de clients tenen saldos baixos, hi ha un nombre considerable d'aquests amb saldos significativament més alts que la resta, arribant fins als 20.000.

Aquesta presència d'outliers explica per què la mitjana del saldo era molt superior a la mediana en les estadístiques descriptives.

####e. Crea el codi corresponent per a cada implementació i mesura el temps d'execució de la creació de la variable. Explica el perquè de les diferències de temps entre ambdues execucions.

####f. Representa un gràfic bivariant de les variables: import de compres, import de compra realitzat en una sola vegada (onoff) i freqüència amb la qual es realitzen les compres. A les opciones de la funció de visualització, indica que no es representi la diagonal de la matriu.

####g. Calcula la matriu de correlació del dataset. Què ens indica?

####h. Quines variables tenen outliers i en quin volum?
Representa la següent informació:

i. Boxplot per a la distribució del núm. de targetes de crèdit per límit establert de la targeta.

ii. Boxplot per a la distribució del núm. de targetes de crèdit en funció de la freqüència de l'actualització del saldo en compte.

Justifica si és necessari eliminar els outliers o tractar-los com a valors extrems perquè poden contenir informació important per al clustering.

####i. Quines són les variables que tenen valors nuls i en quin volum? Hi ha diverses maneres de tractar els valors nuls, quin creus que seria la més correcta en aquest cas: eliminar les observacions o imputar els valors que falten? Heu de valorar i justificar el mètode usat, de ser imputació, useu k-Nearest Neighbors.

####j. Tenen les variables les mateixes unitats o és necessari escalar-les? És necessari estandarditzar les dades si utilitzarem tècniques de clustering? Justifica la teva resposta, si és que sí, escaleu les variables i guardeu les dades en una variable que es digui dades_scaled.

####k. Realitza el PCA a la variable dades_scaled. Sobre la base de la relació de variància explicada per cada component, justifica el núm. raonable de components principals per a obtenir una millor solució en el clustering i guarda’ls en dades_pca format dataframe.

####l. Realitza una reducció de dimensionalitat de 2 components usant l’algoritme umap partint de dades_scaled. Crea una gràfica de visualització i raona si es veuen clústers.